rest/data/version.go

```diff
@@ -83,6 +87,224 @@ func (vc *DBVersionConnector) RestartVersion(versionId string, caller string) er
+				}
+
+				// add the version data into the last rolled-up version
+				newVersion := restModel.APIVersion{}
+				err := newVersion.BuildFromService(&versionFromDB)
+				if err != nil {
+					return nil, errors.Wrapf(err, "error converting version %s from DB model", versionFromDB.Id)
+				}
+				lastRolledUpVersion.Versions = append(lastRolledUpVersion.Versions, newVersion)
+
+				// move on to the next version
+				continue
+			}
+
+			// add a pending rolled-up version, if it exists
+			if lastRolledUpVersion != nil {
+				finalVersions = append(finalVersions, *lastRolledUpVersion)
+				lastRolledUpVersion = nil
+			}
+
+			// if we have hit enough versions, break out
+			if len(finalVersions) == numVersionElements {
+				break
+			}
+
+			// if the version can not be rolled up, create a fully fledged
+			// version for it
+			activeVersion := restModel.APIVersion{}
+			err := activeVersion.BuildFromService(&versionFromDB)
+			if err != nil {
+				return nil, errors.Wrapf(err, "error converting version %s from DB model", versionFromDB.Id)
+			}
+
+			// add the builds to the "row"
+			for _, b := range buildsInVersion {
+				currentRow := buildList[b.BuildVariant]
+				buildsForRow := restModel.APIBuild{}
+				err = buildsForRow.BuildFromService(b)
+				if err != nil {
+					return nil, errors.Wrapf(err, "error converting build %s from DB model", b.Id)
+				}
+
+				currentRow.Builds[versionFromDB.Id] = buildsForRow
+				buildList[b.BuildVariant] = currentRow
+				for _, task := range buildsForRow.TaskCache {
+					if task.Status == evergreen.TaskFailed || task.Status == evergreen.TaskStarted {
+						failedAndStartedTaskIds = append(failedAndStartedTaskIds, task.Id)
+					}
+				}
+			}
+
+			// add the version
+			finalVersions = append(finalVersions, restModel.APIVersions{RolledUp: false, Versions: []restModel.APIVersion{activeVersion}})
+
+		}
+
+		failedAndStartedTasks, err := task.Find(task.ByIds(failedAndStartedTaskIds))
+		if err != nil {
+			return nil, errors.Wrap(err, "error fetching failed tasks")
+
+		}
+
+		for i := range failedAndStartedTasks {
+			if err := failedAndStartedTasks[i].MergeNewTestResults(); err != nil {
+				return nil, errors.Wrap(err, "error merging test results")
+			}
+		}
+
+		addFailedAndStartedTests(buildList, failedAndStartedTasks)
+	}
+
+	// if the last version was rolled-up, add it
+	if lastRolledUpVersion != nil {
+		finalVersions = append(finalVersions, *lastRolledUpVersion)
+	}
+
+	// create the list of display names for the build variants represented
+	buildVariants := []string{}
+	for name := range bvSet {
+		displayName := buildVariantMappings[name]
+		if displayName == "" {
+			displayName = name + " (removed)"
+		}
+		buildVariants = append(buildVariants, displayName)
+	}
+
+	return &restModel.VersionVariantData{
+		Rows:          buildList,
+		Versions:      finalVersions,
+		BuildVariants: buildVariants,
+	}, nil
+
+}
```

Comment:
I really don't want to grumble, but this function length exceeds any sane limit for lines count. And again, I know this was copied. It shouln't block the merge.

Reply:
It's definitely long, but I don't think it's terrible (a lot of the line count is whitespace and comments). I've cleaned it up a bit and simplified some things